overlord/assertstate/assertmgr.go

```diff
@@ -189,10 +202,79 @@ func doFetchSnapAssertions(t *state.Task, _ *tomb.Tomb) error {
 		PrimaryKey: []string{sha3_384},
 	}
 
-	return fetch(t.State(), ref, ss.UserID)
+	err = fetch(t.State(), ref, ss.UserID)
+	if notFound, ok := err.(*assertionNotFoundError); ok {
+		if notFound.ref.Type == asserts.SnapRevisionType {
+			return fmt.Errorf("cannot verify snap %q and its hash, no matching assertions found", ss.Name())
+		} else {
+			return fmt.Errorf("cannot find assertions to verify snap %q and its hash (%v)", ss.Name(), notFound)
+		}
+	}
+	if err != nil {
+		return err
+	}
+
+	err = crossCheckSnap(t.State(), ss.Name(), sha3_384, snapSize, ss.SideInfo)
+	if err != nil {
+		return err
+	}
+
+	// TODO: set DeveloperID from assertions
+	return nil
 }
 
-func undoFetchSnapAssertions(t *state.Task, _ *tomb.Tomb) error {
+func undoFetchCheckSnapAssertions(t *state.Task, _ *tomb.Tomb) error {
 	// nothing to do, the assertions that were *actually* added are still true
 	return nil
 }
+
+func crossCheckSnap(st *state.State, name, snapSHA3_384 string, snapSize uint64, si *snap.SideInfo) error {
+	// get relevant assertions and and do cross checks
+	db := DB(st)
+
+	a, err := db.Find(asserts.SnapRevisionType, map[string]string{
+		"snap-sha3-384": snapSHA3_384,
+	})
+	if err != nil {
+		return fmt.Errorf("internal error: cannot find just fetched snap-revision assertion for %q: %s", name, snapSHA3_384)
+	}
+	snapRev := a.(*asserts.SnapRevision)
+
+	if snapRev.SnapSize() != snapSize {
+		return fmt.Errorf("snap %q file does not have expected size according to assertions (download is broken or tampered): %d != %d", name, snapSize, snapRev.SnapSize())
+	}
+
+	snapID := si.SnapID
+
+	if snapRev.SnapID() != snapID || snapRev.SnapRevision() != si.Revision.N {
+		// we have at least 2 cases here, what's the best message?
+		// - an unsuccesufl MITM
+		// - broken store metadata resulting into broken assertions
+		//   (more likely if it is snap-revision not matching)
+		//   people would need to report this
+		return fmt.Errorf("snap %q file hash %q corresponding assertions implied snap id %q and revision %d are not the ones expected for installing (store metadata is broken or communication tampered): %q and %s", name, snapSHA3_384, snapRev.SnapID(), snapRev.SnapRevision(), snapID, si.Revision)
+	}
+
+	a, err = db.Find(asserts.SnapDeclarationType, map[string]string{
+		"series":  release.Series,
+		"snap-id": snapID,
+	})
+	if err != nil {
+		return fmt.Errorf("internal error: cannot find just fetched snap declaration for %q: %s", name, snapID)
+	}
+	snapDecl := a.(*asserts.SnapDeclaration)
+
+	if snapDecl.SnapName() == "" {
+		// TODO: trigger a global sanity check
```

Comment:
Not sure we need those triggers. The system should be able to self-correct over time nevertheless, so we might just wait for the next pass and have a single code path instead of worrying here.